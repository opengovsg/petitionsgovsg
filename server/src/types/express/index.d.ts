declare namespace Express {
  interface User {
    id: string
    type: string
    fullname: string
  }
}
