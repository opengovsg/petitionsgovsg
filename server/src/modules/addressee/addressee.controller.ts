import { StatusCodes } from 'http-status-codes'
import { createLogger } from '../../bootstrap/logging'
import { Message } from '../../types/message-type'
import { ControllerHandler } from '../../types/response-handler'
import { AddresseeDto, AddresseeService } from './addressee.service'

const logger = createLogger(module)

export class AddresseeController {
  private addresseeService: Public<AddresseeService>

  constructor({
    addresseeService,
  }: {
    addresseeService: Public<AddresseeService>
  }) {
    this.addresseeService = addresseeService
  }

  /**
   * Lists all post
   * @query sort Sort by popularity or recent
   * @query size Number of posts to return
   * @query page If size is given, specify which page to return
   * @return 200 with posts and totalItem for pagination
   * @return 500 when database error occurs
   */
  listAddressees: ControllerHandler<
    undefined,
    AddresseeDto[] | Message,
    undefined,
    undefined
  > = async (req, res) => {
    try {
      const data = await this.addresseeService.listAddressees()
      return res.status(StatusCodes.OK).json(data)
    } catch (error) {
      if (error instanceof Error) {
        logger.error({
          message: 'Error while listing addresees',
          meta: {
            function: 'listAddressees',
          },
          error,
        })
        return res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Server Error' })
      }
    }
  }
}
