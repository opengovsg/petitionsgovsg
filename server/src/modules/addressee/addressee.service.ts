import { Addressee } from '~shared/types/base'
import { ModelDef } from '../../types/sequelize'

export type AddresseeDto = Pick<Addressee, 'id' | 'name' | 'shortName'>

export class AddresseeService {
  private Addressee: ModelDef<Addressee>

  constructor({ Addressee }: { Addressee: ModelDef<Addressee> }) {
    this.Addressee = Addressee
  }

  /**
   * Lists all addressees
   */
  listAddressees = async (): Promise<AddresseeDto[]> => {
    const addressees = (await this.Addressee.findAll({
      attributes: ['id', 'name', 'shortName'],
    })) as AddresseeDto[]
    return addressees
  }
}
