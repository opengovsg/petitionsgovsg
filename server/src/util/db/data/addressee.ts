export const mockAddressee = {
  name: 'Ministry of Environment',
  shortName: 'MoE',
}

export const mockAddressees = [
  {
    id: 1,
    name: 'Ministry of Environment',
    shortName: 'MoE',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
    openToPublic: false,
    officerEmails: [],
  },
  {
    id: 2,
    name: 'Ministry of Labour',
    shortName: 'MoL',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
    openToPublic: false,
    officerEmails: [],
  },
]
