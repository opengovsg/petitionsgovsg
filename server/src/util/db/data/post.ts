import { PostStatus } from '~shared/types/base'

export const mockPost = {
  id: '395ac6ec-dea2-4f78-8dc4-1234cafaff87',
  title: 'Basic petition without optional items',
  reason: 'reason for',
  request: 'request',
  status: PostStatus.Open,
  fullname: 'John Doe',
  salt: '$2b$10$72TKINiR4.ZMaJZmuN5tW.',
  hashedUserSgid: 'wxyz',
  email: 'John@email.com',
  addresseeId: 1,
  summary: null,
  references: null,
  profile: null,
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}

export const mockPosts = [
  {
    id: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    title: 'Basic petition without optional items',
    reason: 'reason for',
    request: 'request',
    status: PostStatus.Open,
    fullname: 'John Doe',
    salt: '$2b$10$72TKINiR4.ZMaJZmuN5tW.',
    hashedUserSgid:
      '$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW',
    email: 'John@email.com',
    addresseeId: 1,
    summary: null,
    references: null,
    profile: null,
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 'aaaaaaaa-dea2-4f78-8dc4-5542cafaff87',
    title: 'Petition with summary, references, profile',
    reason: 'reason',
    request: 'request',
    status: PostStatus.Open,
    fullname: 'John Doe',
    salt: '$2b$10$72TKINiR4.ZMaJZmuN5tX.',
    hashedUserSgid: 'wxyz',
    email: 'John@email.com',
    addresseeId: 1,
    summary: 'petition summary',
    references: 'wikipedia.com',
    profile: 'OGP',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
]
