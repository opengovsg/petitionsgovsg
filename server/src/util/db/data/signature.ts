export const mockSignature = {
  id: 4,
  hashedUserSgid:
    '$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW',
  fullname: null,
  comment: 'I want to support this cause.',
  postId: 1,
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}

export const mockSignatures = [
  {
    id: 1,
    hashedUserSgid: '1234',
    fullname: null,
    comment: null,
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 2,
    hashedUserSgid: '5678',
    fullname: 'John Doe',
    comment: null,
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 3,
    hashedUserSgid: '8765',
    fullname: null,
    comment: 'I support this cause!',
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
]
