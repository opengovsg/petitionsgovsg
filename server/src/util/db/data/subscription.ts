export const mockSubscription = {
  id: 1,
  email: 'test@email.com',
  postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}
