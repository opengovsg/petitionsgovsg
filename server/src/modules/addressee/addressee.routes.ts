import express from 'express'
import { AddresseeController } from './addressee.controller'

export const routeAddressees = ({
  controller,
}: {
  controller: AddresseeController
}): express.Router => {
  const router = express.Router()
  /**
   * Lists all addressees
   * @route   GET /api/posts
   * @returns 200 with posts
   * @returns 500 when database error occurs
   * @access  Public
   */
  router.get('/', controller.listAddressees)
  return router
}
