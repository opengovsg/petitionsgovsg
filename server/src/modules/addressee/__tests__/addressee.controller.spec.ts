import supertest from 'supertest'
import express from 'express'
import { StatusCodes } from 'http-status-codes'
import {
  createTestDatabase,
  resetAndSetupDb,
  SequelizeWithModels,
} from '../../../util/db/jest-db'
import { errAsync } from 'neverthrow'
import { DatabaseError } from '../../core/core.errors'
import { AddresseeController } from '../addressee.controller'
import { mockAddressees } from '../../../util/db/data/addressee'

describe('AddresseeController', () => {
  let db: SequelizeWithModels

  // Mock external service(s)
  const addresseeService = {
    listAddressees: jest.fn(),
  }

  const addresseeController = new AddresseeController({ addresseeService })

  beforeAll(async () => {
    db = await createTestDatabase()
  })

  beforeEach(async () => {
    await resetAndSetupDb(db)
    jest.resetAllMocks()
  })

  describe('listAddressees', () => {
    const app = express()
    app.get('/', addresseeController.listAddressees)
    const request = supertest(app)
    it('retrieves all addressees', async () => {
      addresseeService.listAddressees.mockResolvedValueOnce(mockAddressees)

      const response = await request.get('/')

      expect(addresseeService.listAddressees).toHaveBeenCalled()
      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual(mockAddressees)
      expect(response.body).toMatchSnapshot()
    })

    it('returns INTERNAL_SERVER_ERROR on bad service', async () => {
      addresseeService.listAddressees.mockRejectedValue(
        errAsync(new DatabaseError()),
      )

      const response = await request.get('/')

      expect(addresseeService.listAddressees).toHaveBeenCalled()
      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({ message: 'Server Error' })
      expect(response.body).toMatchSnapshot()
    })
  })
})
