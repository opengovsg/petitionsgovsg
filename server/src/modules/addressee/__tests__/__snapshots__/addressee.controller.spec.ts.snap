// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddresseeController listAddressees retrieves all addressees 1`] = `
Array [
  Object {
    "createdAt": "2022-01-01T00:00:00+00:00",
    "id": 1,
    "name": "Ministry of Environment",
    "officerEmails": Array [],
    "openToPublic": false,
    "shortName": "MoE",
    "updatedAt": "2022-01-01T00:00:00+00:00",
  },
  Object {
    "createdAt": "2022-01-01T00:00:00+00:00",
    "id": 2,
    "name": "Ministry of Labour",
    "officerEmails": Array [],
    "openToPublic": false,
    "shortName": "MoL",
    "updatedAt": "2022-01-01T00:00:00+00:00",
  },
]
`;

exports[`AddresseeController listAddressees returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;
